src/metabase/db/metadata_queries.clj

```diff
@@ -36,39 +36,47 @@
-                            (if (some? existing-filter)
-                              [:and existing-filter filter-form]
-                              filter-form)))))
+(defn add-required-filter-if-needed
+  "Add a dummy filter for tables that require a filters.
+  Look into tables from source tables and all the joins.
+  Currently this only apply to partitioned tables on bigquery that requires a partition filter."
+  [query]
+  (let [table-ids              (->> (conj (keep :source-table (:joins query)) (:source-table query))
+                                    (filter pos-int?))
+        required-filter-fields (when (seq table-ids)
+                                 (t2/select :model/Field {:select    [:f.*]
+                                                          :from      [[:metabase_field :f]]
+                                                          :left-join [[:metabase_table :t] [:= :t.id :f.table_id]]
+                                                          :where     [:and
+                                                                      [:= :f.active true]
+                                                                      [:= :f.database_partitioned true]
+                                                                      [:= :t.active true]
+                                                                      [:= :t.database_require_filter true]
+                                                                      [:in :t.id table-ids]]}))
+        update-query-filter-fn (fn [existing-filter new-filter]
+                                (if (some? existing-filter)
+                                  [:and existing-filter new-filter]
+                                  new-filter))]
+    (case (count required-filter-fields)
+      0
+      query
+      1
+      (update query :filter update-query-filter-fn (partition-field->filter-form (first required-filter-fields)))
+      ;; > 1
+      (update query :filter update-query-filter-fn (into [:and] (map partition-field->filter-form required-filter-fields))))))
 
 (defn- field-mbql-query
   [table mbql-query]
-  (cond-> mbql-query
-    true
-    (assoc :source-table (:id table))
-
-    ;; Some table requires a filter to be able to query the data
-    ;; Currently this only applied to Partitioned table in bigquery where the partition field
-    ;; is required as a filter.
-    ;; In the future we probably want this to be dispatched by database engine type
-    (:database_require_filter table)
-    (query-with-default-partitioned-field-filter (:id table))))
-
-(defn- field-query [{table-id :table_id} mbql-query]
+  (-> mbql-query
+      (assoc :source-table (:id table))
+      ;; Some table requires a filter to be able to query the data
+      ;; Currently this only applied to Partitioned table in bigquery where the partition field
+      ;; is required as a filter.
+      ;; In the future we probably want this to be dispatched by database engine type
```

Comment:
Since you've moved the logic for whether we need a filter or not into `add-required-filter-if-needed`, I think this comment could just be part of its docstring. You've duplicated some of what it says already, so I'd suggest DRYing it up and removing this comment so it doesn't risk getting out of date